c10/core/Scalar.h

```diff
@@ -186,6 +187,29 @@ class C10_API Scalar {
   } v;
 };
 
+struct OptionalScalarRef {
+  OptionalScalarRef() : scalar_(nullptr) {}
+  OptionalScalarRef(const Scalar* scalar) : scalar_(scalar) {
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(scalar_);
+  }
+
+  bool has_value() const {
+    return scalar_ != nullptr;
+  }
+
+  const Scalar& toScalar() const {
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(scalar_);
+    return *scalar_;
+  }
+
+  operator bool() const {
+    return has_value();
+  }
```

Comment:
btw, if you hate optional's conversion to boolean, you could delete this implicit conversion and force people to write `has_value()` explicitly.